src/generator/extract_schema.cpp

```diff
@@ -0,0 +1,323 @@
+    }
+    ddwaf_object_array_add(&array, &meta);
+
+    return array;
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+ddwaf_object node_serialize::operator()(const node_record_ptr &node) const noexcept
+{
+    ddwaf_object tmp;
+    ddwaf_object array;
+    ddwaf_object_array(&array);
+
+    ddwaf_object map;
+    ddwaf_object_map(&map);
+    for (const auto &[key, child] : node->children) {
+        auto res = std::visit(node_serialize{}, child);
+        ddwaf_object_map_addl(&map, key.data(), key.size(), &res);
+    }
+    ddwaf_object_array_add(&array, &map);
+
+    if (node->truncated) {
+        ddwaf_object meta;
+        ddwaf_object_map(&meta);
+        ddwaf_object_map_add(&meta, "truncated", ddwaf_object_bool(&tmp, true));
+        ddwaf_object_array_add(&array, &meta);
+    }
+
+    return array;
+}
+
+ddwaf_object serialize(const base_node &node) { return std::visit(node_serialize{}, node); }
+
+// NOLINTNEXTLINE(misc-no-recursion)
+base_node generate_helper(const ddwaf_object *object, std::size_t depth)
+{
+    switch (object->type) {
+    case DDWAF_OBJ_NULL:
+        return node_scalar{scalar_type::null};
+    case DDWAF_OBJ_FLOAT:
+        return node_scalar{scalar_type::real};
+    case DDWAF_OBJ_BOOL:
+        return node_scalar{scalar_type::boolean};
+    case DDWAF_OBJ_STRING:
+        return node_scalar{scalar_type::string};
+    case DDWAF_OBJ_SIGNED:
+    case DDWAF_OBJ_UNSIGNED:
+        return node_scalar{scalar_type::integer};
+    case DDWAF_OBJ_MAP: {
+        auto length = static_cast<std::size_t>(object->nbEntries);
+        node_record_ptr record = std::make_unique<node_record>();
+        if (length > extract_schema::max_record_nodes) {
+            record->truncated = true;
+            length = extract_schema::max_record_nodes;
+        }
+        for (std::size_t i = 0; i < length && depth > 1; i++) {
+            const auto *child = &object->array[i];
+            if (child->parameterName == nullptr) {
+                continue;
+            }
+
+            auto schema = generate_helper(child, depth - 1);
+            std::string_view key{
+                child->parameterName, static_cast<std::size_t>(child->parameterNameLength)};
+            record->children.emplace(key, std::move(schema));
+        }
+        return record;
+    }
+    case DDWAF_OBJ_ARRAY: {
+        auto length = static_cast<std::size_t>(object->nbEntries);
+        node_array_ptr array = std::make_unique<node_array>();
+        array->length = length;
+        if (length > extract_schema::max_array_nodes) {
+            array->truncated = true;
+            length = extract_schema::max_array_nodes;
+        }
+        for (std::size_t i = 0; i < length && depth > 1; i++) {
+            const auto *child = &object->array[i];
+            auto schema = generate_helper(child, depth - 1);
+            array->children.emplace(std::move(schema));
+        }
+        return array;
+    }
```

Comment:
I'm a bit perplexed with this separate logic for maps and arrays. So if my original data is `Map<A, B>`, then the structure of B will be repeated for every encountered value of A. If OTOH I have a `List<B>`, then there is this expensive effort, with equality/hash recursively checking for equality, to deduplicate `B`.